src/ray/common/asio/instrumented_io_context.h

```diff
@@ -28,7 +28,45 @@ class instrumented_io_context : public boost::asio::io_context {
  public:
   /// Initializes the global stats struct after calling the base contructor.
   /// TODO(ekl) allow taking an externally defined event tracker.
-  instrumented_io_context() : event_stats_(std::make_shared<EventTracker>()) {}
+  instrumented_io_context()
+      : is_running_{false}, event_stats_(std::make_shared<EventTracker>()) {}
+
+  /// Run the io_context if and only if no other thread is running it. Blocks
+  /// other threads from running once started. Noop if there is a thread
+  /// running it already.
+  void run_if_not_running(std::function<void()> prerun_fn) {
```

Comment:
Let's add comment for this one about the actual use case and mentioning that it's only used in GcsClient::Connect. It'll be deprecated later.